### ISSUE TITLE
MAVGEN LUA WIRESHARK cannot handle negative lattitudes/longitudes

### ISSUE THREAD

**Issue by PeterJBurke** [mavlink/mavlink#2073]
I have created the LUA script to work with Wireshark to parse mavlink UDP packets, 
using the directions here:
https://mavlink.io/en/guide/wireshark.html
Specifically, I run the command: 
python3 -m pymavlink.tools.mavgen --lang=WLua --wire-protocol=2.0 --output=mavlink_2_common message_definitions/v1.0/common.xml
It creates the lua script "mavlink_2_common.lua" which I then incorporate with Wireshark.
It all works fine, but when there is a negative latitude or longitude, it shows the bits correctly but a zero for the values.
The screenshot shows the 4 bytes for the lat in blue are correctly displayed as the lat , but the next four bytes show zero for the lat above. (The actual lat is -117... if you check the 4 bytes).
So it must be something about signed vs unsigned int32 in Lua/Mavlink.
<img width="1233" alt="Screenshot 2024-01-11 at 7 46 18 PM" src="https://github.com/mavlink/mavlink/assets/42294224/7cfaf02b-d142-453f-a7ea-e16a1190cd6d">

**Comment by shancock884** [comment]
This sounds like it might be the same issue as #1862, which was cross-posted to https://github.com/ArduPilot/pymavlink/issues/701, as generators are part of pymavlink module.

Reading those previous issues, it seems that no-one got to the bottom of it at the time, but one hypothesis was that this was specific to Mac.... can you confirm what platform you are using (your screenshot certainly looks like a Mac!), and also what version of pymavlink (i.e. mavgen) and Wireshark you are using.

With Wireshark v4.2 and pymavlink 2.4.41, I can certainly see negative values on my Linux system.

**Comment by shancock884** [comment]
@rotu It looked from the screenshot in https://github.com/mavlink/mavlink/issues/2105 that you are on Mac, right? I assume you can see negative lat/lon values ok?!

**Comment by rotu** [comment]
> @rotu It looked from the screenshot in https://github.com/mavlink/mavlink/issues/2105 that you are on Mac, right? I assume you can see negative lat/lon values ok?!

No. This appears to be a regression and I can see the impact on Mac in the tests as well:

```console
pytest -vv
...
tests/snapshottests/test_wlua.py::test_wlua[common.xml-gps_global_origin.pcapng] FAILED       [ 98%]
tests/snapshottests/test_wlua.py::test_wlua[common.xml-mission_item_int.pcapng] FAILED        [100%]                     
...
__________________________ test_wlua[common.xml-gps_global_origin.pcapng] ___________________________
...
E       assert [+ received] == [- snapshot]
E           dict({
E                 ...
E                       latitude (int32_t) [degE7]: 517015586 (51.7015586 deg)
E         -             longitude (int32_t) [degE7]: -3332886 (-0.3332886 deg)
E         +             longitude (int32_t) [degE7]: 0 (-0.3332886 deg)
E                       altitude (int32_t) [mm]: 56990
E                 ...
E           })
...
___________________________ test_wlua[common.xml-mission_item_int.pcapng] ___________________________
...
>       assert props_to_match == snapshot()
E       assert [+ received] == [- snapshot]
E           dict({
E                 ...
E                       x: Latitude (int32_t): 514775152
E         -             y: Longitude (int32_t): -25939161
E         +             y: Longitude (int32_t): 0
E                       z: Ground Altitude (float) m: 0
E                 ...
E           })


```

**Comment by shancock884** [comment]
Ok thanks for checking that.... Very odd!
Not sure if its a regression or always been like that on mac, as https://github.com/mavlink/mavlink/issues/1862 first mentions it nearly 2 years ago.
I can try to make a minimum LUA+PCAP combo to try out on Windows/Linux/Mac, and if need-be will create an issue on Wireshark's Gitlab.

**Comment by rotu** [comment]
Guessing it's this line: https://gitlab.com/wireshark/wireshark/-/blob/259bbbff2a5257319ff69aa32cc377fc501eb022/epan/wslua/wslua_tree.c#L535

```c
case FT_INT32:
  item = proto_tree_add_int(tree_item->tree,hfid,tvbr->tvb->ws_tvb,tvbr->offset,tvbr->len,wslua_checkuint32(L,1));
  break;
```

And that it's supposed to be `wslua_checkint32` instead of `wslua_checkuint32`.

**Comment by shancock884** [comment]
Good spot - yes that looks like it could be the culprit!
If I am following it correctly, the incoming number gets cast to an uint32, then this is passed as an argument to a function expecting an int32.
I guess that somewhere in this lot, the behaviour between Mac and Linux differs, and the negative number gets cut at 0.
Maybe the equivalent of something like this (which prints -100 on Linux):
```
#include <stdio.h>
void func(int v) {
  printf("%d\n",v);
}
int main(void) {
  double x = -100;
  unsigned int y = (unsigned int)x;
  func(y);
  return 0;
}
```

In the meantime, a fix on pymavlink side would be to remove the value argument from calls to tree:add_le, but we do need to be a little careful with this, as there are cases where we need it. For example when a float command parameter represents an enum, we need to use the value, as otherwise Wireshark would incorrectly interpret the float bytes as a raw uint32.

**Comment by rotu** [comment]
> I guess that somewhere in this lot, the behaviour between Mac and Linux differs, and the negative number gets cut at 0.
> Maybe the equivalent of something like this (which prints -100 on Linux):
> ```c
>   double x = -100;
>   unsigned int y = (unsigned int)x;
> ```

Honestly, I thought that was a ridiculous suggestion and there's no way that differs between Linux and Mac. ...but it turns out you're right! It does indeed print 0 on my computer when compiling with both clang 15 and gcc 13!

> In the meantime, a fix on pymavlink side would be to remove the value argument from calls to tree:add_le, but we do need to be a little careful with this, as there are cases where we need it. For example when a float command parameter represents an enum, we need to use the value, as otherwise Wireshark would incorrectly interpret the float bytes as a raw uint32.

It might also make sense to use `ftypes.DOUBLE` for numeric values but I think that changes the semantics of `valuestring` in such a way that it no longer describes enum values.

BTW submitted merge request at https://gitlab.com/wireshark/wireshark/-/merge_requests/15438

**Comment by shancock884** [comment]
I had a crack at a quick workaround/fix on the pymavlink side in this branch: https://github.com/shancock884/pymavlink/tree/mac-neg-int-fix
It removes the 'value' arg from calls to add_le, unless there is an associated enum - to make sure it doesn't break the case of a double or float being linked to an enum, as is the case on a number of command messages.
Its not very optimised yet, as 'value = tvbrange:...' line is still generated, even when not needed, but should prove the concept.

**Comment by rotu** [comment]
It would also suffice to promote integers to int64.

```lua
f.cmd_MAV_CMD_NAV_VTOL_LAND_y = ProtoField.new("y: Longitude (int32_t)", "mavlink_proto.cmd_MAV_CMD_NAV_VTOL_LAND_y", ftypes.INT64, nil)

-- ...

    value = tvbrange:le_int64()
    subtree = tree:add_le(f.cmd_MAV_CMD_NAV_VTOL_LAND_y, tvbrange, value)
```

That said, I don't like this code at all :-(.
